Refactor Data Input/Output Operators

jortiz16 opened the pull request:
I renamed some operators as we discussed on Friday. I know this is still tentative, but it can be easily changed if we vote against it on Monday:

```DataInput``` --> ```TupleSource```
```DataOutput``` --> ```TupleSink```
```TupleSource``` --> ```BatchTupleSource``` (this class is usually used for testing, I wasn't sure what to call it but I was forced to rename it so it would not conflict with the new TupleSource class above)
```SinkRoot``` --> ```EmptySink```

This PR creates the new TupleSource operator, which eliminates the need for the FileScan and BinaryScan. Corresponding TupleReaders are created for these classes (CsvTupleReader and BinaryTupleReader). The other scans left are the Tipsy, NChilada and SeaFlow scans.

senderista commented on src/edu/washington/escience/myria/TupleReader.java:
```diff
@@ -0,0 +1,30 @@
+
+  TupleBatch readTuples() throws IOException, DbException;
+
+  void done() throws IOException;
```
Any reason not to call this `close()` for symmetry with `open()`?

senderista commented on src/edu/washington/escience/myria/api/encoding/QueryConstruct.java:
```diff
@@ -313,7 +313,7 @@ private static void verifyAndPropagateLocalEdgeConstraints(final List<PlanFragme
   /**
    * Use the Catalog to set the workers for fragments that have scans, and verify that the workers are consistent with
-   * existing constraints.
+   * existing constraints. f
```
typo

senderista commented on src/edu/washington/escience/myria/operator/DataInput.java:
```diff
@@ -0,0 +1,55 @@
+  }
+
+  @Override
+  protected void init(final ImmutableMap<String, Object> execEnvVars) throws DbException {
```
Don't we also need to `close()` `tupleReader` in a `cleanup()` override?

senderista commented on systemtest/edu/washington/escience/myria/systemtest/JsonOperatorTests.java:
```diff
@@ -122,6 +123,6 @@ public void splitTest() throws Exception {
     String expectedData =
         "[{\"string_array\":\"a:b:c:d:e:f\",\"string_array_splits\":\"a\"},{\"string_array\":\"a:b:c:d:e:f\",\"string_array_splits\":\"b\"},{\"string_array\":\"a:b:c:d:e:f\",\"string_array_splits\":\"c\"},{\"string_array\":\"a:b:c:d:e:f\",\"string_array_splits\":\"d\"},{\"string_array\":\"a:b:c:d:e:f\",\"string_array_splits\":\"e\"},{\"string_array\":\"a:b:c:d:e:f\",\"string_array_splits\":\"f\"} ...
-    assertEquals(expectedData, data);
+    assertEquals(data, expectedData);
```
The original order of args is correct: http://junit.sourceforge.net/javadoc/org/junit/Assert.html#assertEquals(java.lang.Object, java.lang.Object)

senderista commented on src/edu/washington/escience/myria/BinaryTupleReader.java:
```diff
@@ -41,32 +32,48 @@
   private transient DataInput dataInput;
 
   /**
+   * Construct a new BinaryFileScan object that reads the given binary file and creates tuples from the file data that
```
`BinaryTupleReader`

senderista commented on src/edu/washington/escience/myria/BinaryTupleReader.java:
```diff
@@ -41,32 +32,48 @@
+  public void open(final InputStream stream) throws IOException, DbException {
+    buffer = new TupleBatchBuffer(getSchema());
+    InputStream inputStream;
+    inputStream = new BufferedInputStream(stream);
```
collapse into one line

senderista commented on src/edu/washington/escience/myria/BinaryTupleReader.java:
```diff
@@ -41,32 +32,48 @@
+  }
+
+  /**
    * Construct a new BinaryFileScan object that reads the given binary file and create tuples from the file data that
```
`BinaryTupleReader`

senderista commented:
LGTM, we already reached consensus on renaming so I think this is ready to merge.

jortiz16 commented:
I also ended up running a rebase + force push for this PR. The only new relevant commit is the "fixing comments based on code review". I fixed the "BinaryFileScan" comments and I also fixed the first typo you described but I fixed it in an earlier commit (you won't see this change in the newest commit).

jortiz16 commented:
oops I missed other comments you had earlier... fixing now...

jortiz16 commented:
ok now it should be good

senderista commented:
I forgot that the corresponding work in raco hasn't been done yet, so users already hit a regression. Need to revert this merge until the raco work is complete.

senderista commented:
Since there's nothing wrong (AFAIK) with these changes, I'll just tag the branch before the merge and use that tag in the Myria deployment script until the raco changes are ready.

jortiz16 commented:
I completely forgot this as well! Thanks for catching this.
